homeassistant/components/zha/config_flow.py

```diff
@@ -53,48 +165,276 @@ async def async_step_user(self, user_input=None):
+
+        if self._current_settings is not None:
+            strategies.append(FORMATION_REUSE_SETTINGS)
+
+        strategies.append(FORMATION_RESTORE_MANUAL_BACKUP)
+        strategies.append(FORMATION_FORM_NEW_NETWORK)
+
+        # Pick a default. Forming a new network will always be an option but we would
+        # prefer to suggest something non-destructive if it's available.
+        suggested_strategy = [
+            strategy
+            for strategy in strategies
+            if strategy
+            in (
+                FORMATION_RESTORE_AUTOMATIC_BACKUP,
+                FORMATION_REUSE_SETTINGS,
+                FORMATION_FORM_NEW_NETWORK,
+            )
+        ][0]
+
+        schema = {
+            vol.Required(FORMATION_STRATEGY, default=suggested_strategy): vol.In(
+                strategies
+            ),
+        }
 
-        schema = {vol.Required(CONF_RADIO_TYPE): vol.In(sorted(RadioType.list()))}
         return self.async_show_form(
-            step_id="pick_radio",
+            step_id="choose_formation_strategy",
             data_schema=vol.Schema(schema),
         )
 
+    async def async_step_upload_manual_backup(self, user_input=None):
+        """Upload and restore a coordinator backup JSON file."""
+        errors = {}
+
+        if user_input is not None:
+            uploaded_file_id = user_input[UPLOADED_BACKUP_FILE]
+
+            with process_uploaded_file(self.hass, uploaded_file_id) as file_path:
+                backup_json = await self.hass.async_add_executor_job(
+                    file_path.read_text
+                )
+
+                try:
+                    backup_obj = json.loads(backup_json)
+                    backup = zigpy.backups.NetworkBackup.from_dict(backup_obj)
+                except ValueError:
+                    errors["base"] = "invalid_backup_json"
+                else:
+                    if user_input.get(OVERWRITE_COORDINATOR_IEEE):
+                        backup = self._allow_overwrite_ezsp_ieee(backup)
+
+                    async with self._connect_zigpy_app() as app:
+                        await app.backups.restore_backup(backup)
+
+                    return await self._async_create_radio_entity()
+
+        data_schema = {
+            vol.Required(UPLOADED_BACKUP_FILE): FileSelector(
+                FileSelectorConfig(accept=".json,application/json")
+            )
+        }
+
+        if self._radio_type == RadioType.ezsp and (
+            self._current_settings is None
+            or (
+                self._current_settings.network_info.metadata["ezsp"][
+                    "can_write_custom_eui64"
+                ]
+            )
+        ):
+            data_schema[vol.Required(OVERWRITE_COORDINATOR_IEEE, default=True)] = bool
+
+        return self.async_show_form(
+            step_id="upload_manual_backup",
+            data_schema=vol.Schema(data_schema),
+            errors=errors,
+        )
+
+    def _format_backup_choice(
```

Comment:
I'd make this a standalone function as it doesn't need to know anything about the flow. It makes it easier to test this util function.